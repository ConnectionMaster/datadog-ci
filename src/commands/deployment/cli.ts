import {DeploymentMarkCommand} from './mark'

module.exports = [DeploymentMarkCommand]