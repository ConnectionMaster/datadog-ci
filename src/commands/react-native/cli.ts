import {UploadCommand} from './upload'
import {XCodeCommand} from './xcode'

module.exports = [UploadCommand, XCodeCommand]
