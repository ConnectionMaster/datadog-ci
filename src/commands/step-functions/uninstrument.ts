import {CloudWatchLogs, StepFunctions} from 'aws-sdk'
import {Command} from 'clipanion'

import {deleteSubscriptionFilter, describeStateMachine, describeSubscriptionFilters, untagResource} from './aws'
import {displayChanges, applyChanges} from './changes'
import {TAG_VERSION_NAME} from './constants'
import {getStepFunctionLogGroupArn, isValidArn, parseArn} from './helpers'
import {RequestsByStepFunction} from './interfaces'

export class UninstrumentStepFunctionsCommand extends Command {
  public static usage = Command.Usage({
    description: 'Unubscribe Step Function Log Groups from a Datadog Forwarder',
    details: '--step-function expects a Step Function ARN\n--forwarder expects a Lambda ARN',
    examples: [
      [
        'View and apply changes to unsubscribe a Step Function Log Group from a Datadog Forwarder',
        'datadog-ci step-functions uninstrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction --forwarder arn:aws:lambda:us-east-1:000000000000:function:ExampleDatadogForwarder',
      ],
      [
        'View changes to unsubscribe a Step Function Log Group from a Datadog Forwarder',
        'datadog-ci step-functions uninstrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction --forwarder arn:aws:lambda:us-east-1:000000000000:function:ExampleDatadogForwarder --dry-run',
      ],
      [
        'View and apply changes to unsubscribe multiple Step Function Log Groups from a Datadog Forwarder',
        'datadog-ci step-functions uninstrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction1 --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction2 --forwarder arn:aws:lambda:us-east-1:000000000000:function:ExampleDatadogForwarder',
      ],
    ],
  })

  private dryRun = false
  private forwarderArn!: string
  private stepFunctionArns: string[] = []

  public async execute() {
    let validationError = false
    if (typeof this.forwarderArn !== 'string') {
      this.context.stdout.write('[Error] --forwarder is required\n')
      validationError = true
    } else if (!isValidArn(this.forwarderArn)) {
      this.context.stdout.write(`[Error] invalid arn format for --forwarder ${this.forwarderArn}\n`)
      validationError = true
    }

    // remove duplicate step function arns
    const stepFunctionArns = [...new Set(this.stepFunctionArns)]

    if (stepFunctionArns.length === 0) {
      this.context.stdout.write(`[Error] must specify at least one --step-function\n`)
      validationError = true
    }

    for (const stepFunctionArn of stepFunctionArns) {
      if (!isValidArn(stepFunctionArn)) {
        this.context.stdout.write(`[Error] invalid arn format for --step-function ${stepFunctionArn}\n`)
        validationError = true
      }
    }

    if (validationError) {
      return 1
    }

    const requestsByStepFunction: RequestsByStepFunction = {}

    // loop over step functions passed as parameters and generate a list of requests to make to AWS for each step function
    for (const stepFunctionArn of this.stepFunctionArns) {
      requestsByStepFunction[stepFunctionArn] = []

      // use region from the step function arn to make requests to AWS
      const arnObject = parseArn(stepFunctionArn)
      const region = arnObject.region
      const cloudWatchLogsClient = new CloudWatchLogs({region})
      const stepFunctionsClient = new StepFunctions({region})

      let stepFunction
      try {
        stepFunction = await describeStateMachine(stepFunctionsClient, stepFunctionArn)
      } catch (err) {
        if (err instanceof Error) {
          this.context.stdout.write(`\n[Error] ${err.message}. Unable to fetch Step Function ${stepFunctionArn}\n`)
        } else {
          this.context.stdout.write(`\n[Error] ${err}. Unable to fetch Step Function ${stepFunctionArn}\n`)
        }

        return 1
      }

      // the log group that should be unsubscribed from the forwarder is parsed from the step function logging config
      const logGroupArn = getStepFunctionLogGroupArn(stepFunction)
      if (logGroupArn === undefined) {
        this.context.stdout.write('\n[Error] Unable to get log group arn from Step Function logging configuration\n')

        return 1
      }
      const logGroupName = parseArn(logGroupArn).resourceName

      // delete subscription filters that are subscribed to the specified forwarder
      let describeSubscriptionFiltersResponse: CloudWatchLogs.DescribeSubscriptionFiltersResponse | undefined
      try {
        describeSubscriptionFiltersResponse = await describeSubscriptionFilters(cloudWatchLogsClient, logGroupName)
      } catch (err) {
        if (err instanceof Error) {
          this.context.stdout.write(
            `\n[Error] ${err.message}. Unable to fetch subscription filter to delete for Log Group ${logGroupName}\n`
          )
        } else {
          this.context.stdout.write(
            `\n[Error] ${err}. Unable to fetch subscription filter to delete for Log Group ${logGroupName}\n`
          )
        }

        return 1
      }

      const subscriptionFilters =
        describeSubscriptionFiltersResponse.subscriptionFilters?.filter(
          (subscriptionFilter) => subscriptionFilter.destinationArn === this.forwarderArn
        ) ?? []

      for (const subscriptionFilter of subscriptionFilters) {
        if (typeof subscriptionFilter.filterName === 'string') {
          const deleteSubscriptionFilterRequest = deleteSubscriptionFilter(
            cloudWatchLogsClient,
            subscriptionFilter.filterName,
            logGroupName
          )
          requestsByStepFunction[stepFunctionArn].push(deleteSubscriptionFilterRequest)
        }
      }

      const tagKeystoRemove: StepFunctions.TagKeyList = [TAG_VERSION_NAME]
      const untagStepFunctionRequest = untagResource(stepFunctionsClient, stepFunctionArn, tagKeystoRemove)
      requestsByStepFunction[stepFunctionArn].push(untagStepFunctionRequest)
    }

    if (Object.keys(requestsByStepFunction).length === 0) {
      this.context.stdout.write('No changes to apply')

      return 0
    }

    // display changes that will be applied if dry run mode is disabled
    displayChanges(requestsByStepFunction, this.dryRun, this.context)

    // if dry run mode is disabled, apply changes by making requests to AWS
    if (!this.dryRun) {
      const error = await applyChanges(requestsByStepFunction, this.context)
      if (error) {
        return 1
      }
    }

    return 0
  }
}

UninstrumentStepFunctionsCommand.addPath('step-functions', 'uninstrument')

UninstrumentStepFunctionsCommand.addOption('dryRun', Command.Boolean('-d,--dry-run'))
UninstrumentStepFunctionsCommand.addOption('forwarderArn', Command.String('--forwarder'))
UninstrumentStepFunctionsCommand.addOption('stepFunctionArns', Command.Array('-s,--step-function'))
