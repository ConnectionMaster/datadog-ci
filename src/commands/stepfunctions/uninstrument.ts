import {CloudWatchLogs, StepFunctions} from 'aws-sdk'
import {Command} from 'clipanion'

import {deleteSubscriptionFilter, describeStateMachine, describeSubscriptionFilters, untagResource} from './aws'
import {displayChanges, applyChanges} from './changes'
import {TAG_VERSION_NAME} from './constants'
import {DD_CI_IDENTIFING_STRING, getStepFunctionLogGroupArn, isValidArn, parseArn} from './helpers'
import {RequestsByStepFunction} from './interfaces'

export class UninstrumentStepFunctionsCommand extends Command {
  public static usage = Command.Usage({
    description: 'Remove Step Functions log groups subscription filter created by Datadog-CI',
    details: '--stepfunction expects a Step Function ARN',
    examples: [
      [
        'View and apply changes to remove Step Functions log groups subscription filters created by Datadog-CI',
        'datadog-ci stepfunctions uninstrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
      ],
      [
        'View changes to remove Step Functions log groups subscription filters created by Datadog-CI',
        'datadog-ci stepfunctions uninstrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction --dry-run',
      ],
      [
        'View and apply changes to remove Step Functions log groups subscription filters created by Datadog-CI',
        'datadog-ci stepfunctions uninstrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction1 --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction2',
      ],
    ],
  })

  private dryRun = false
  private stepFunctionArns: string[] = []

  public async execute() {
    let validationError = false

    // remove duplicate step function arns
    const stepFunctionArns = [...new Set(this.stepFunctionArns)]

    if (stepFunctionArns.length === 0) {
      this.context.stdout.write(`[Error] must specify at least one --step-function\n`)
      validationError = true
    }

    for (const stepFunctionArn of stepFunctionArns) {
      if (!isValidArn(stepFunctionArn)) {
        this.context.stdout.write(`[Error] invalid arn format for --step-function ${stepFunctionArn}\n`)
        validationError = true
      }
    }

    if (validationError) {
      return 1
    }

    const requestsByStepFunction: RequestsByStepFunction = {}

    // loop over step functions passed as parameters and generate a list of requests to make to AWS for each step function
    for (const stepFunctionArn of stepFunctionArns) {
      requestsByStepFunction[stepFunctionArn] = []

      // use region from the step function arn to make requests to AWS
      const arnObject = parseArn(stepFunctionArn)
      const region = arnObject.region
      const cloudWatchLogsClient = new CloudWatchLogs({region})
      const stepFunctionsClient = new StepFunctions({region})

      let stepFunction
      try {
        stepFunction = await describeStateMachine(stepFunctionsClient, stepFunctionArn)
      } catch (err) {
        if (err instanceof Error) {
          this.context.stdout.write(`\n[Error] ${err.message}. Unable to fetch Step Function ${stepFunctionArn}\n`)
        } else {
          this.context.stdout.write(`\n[Error] ${err}. Unable to fetch Step Function ${stepFunctionArn}\n`)
        }

        return 1
      }

      const logGroupArn = getStepFunctionLogGroupArn(stepFunction)
      if (logGroupArn === undefined) {
        this.context.stdout.write('\n[Error] Unable to get Log Group arn from Step Function logging configuration\n')

        return 1
      }
      const logGroupName = parseArn(logGroupArn).resourceName

      // delete subscription filters that are created by Datadog-CI
      let describeSubscriptionFiltersResponse: CloudWatchLogs.DescribeSubscriptionFiltersResponse | undefined
      try {
        describeSubscriptionFiltersResponse = await describeSubscriptionFilters(cloudWatchLogsClient, logGroupName)
      } catch (err) {
        if (err instanceof Error) {
          this.context.stdout.write(
            `\n[Error] ${err.message}. Unable to fetch Subscription Filter to delete for Log Group ${logGroupName}\n`
          )
        } else {
          this.context.stdout.write(
            `\n[Error] ${err}. Unable to fetch Subscription Filter to delete for Log Group ${logGroupName}\n`
          )
        }

        return 1
      }
      const subscriptionFilters =
        describeSubscriptionFiltersResponse.subscriptionFilters?.filter((subscriptionFilter) =>
          subscriptionFilter.filterName?.includes(DD_CI_IDENTIFING_STRING)
        ) ?? []

      for (const subscriptionFilter of subscriptionFilters) {
        if (typeof subscriptionFilter.filterName === 'string') {
          const deleteSubscriptionFilterRequest = deleteSubscriptionFilter(
            cloudWatchLogsClient,
            subscriptionFilter.filterName,
            logGroupName
          )
          requestsByStepFunction[stepFunctionArn].push(deleteSubscriptionFilterRequest)
        }
      }

      const tagKeystoRemove: StepFunctions.TagKeyList = [TAG_VERSION_NAME]
      const untagStepFunctionRequest = untagResource(stepFunctionsClient, stepFunctionArn, tagKeystoRemove)
      requestsByStepFunction[stepFunctionArn].push(untagStepFunctionRequest)
    }

    if (Object.keys(requestsByStepFunction).length === 0) {
      this.context.stdout.write('No changes to apply')

      return 0
    }

    // display changes that will be applied if dry run mode is disabled
    displayChanges(requestsByStepFunction, this.dryRun, this.context)

    // if dry run mode is disabled, apply changes by making requests to AWS
    if (!this.dryRun) {
      const error = await applyChanges(requestsByStepFunction, this.context)
      if (error) {
        return 1
      }
    }

    return 0
  }
}

UninstrumentStepFunctionsCommand.addPath('stepfunctions', 'uninstrument')

UninstrumentStepFunctionsCommand.addOption('dryRun', Command.Boolean('-d,--dry-run'))
UninstrumentStepFunctionsCommand.addOption('stepFunctionArns', Command.Array('-s,--step-function'))
