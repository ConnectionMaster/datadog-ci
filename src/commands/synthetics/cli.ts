import {RunTestCommand} from './command'

module.exports = [RunTestCommand]
