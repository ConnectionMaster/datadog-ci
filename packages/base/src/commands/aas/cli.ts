import {AasInstrumentCommand} from './instrument'
import {AasUninstrumentCommand} from './uninstrument'

// prettier-ignore
export const commands = [
  AasInstrumentCommand,
  AasUninstrumentCommand,
]
