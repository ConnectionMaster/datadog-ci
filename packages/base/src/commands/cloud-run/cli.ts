/* eslint-disable import-x/order */
import {CloudRunFlareCommand} from './flare'
import {CloudRunInstrumentCommand} from './instrument'
import {CloudRunUninstrumentCommand} from './uninstrument'

// prettier-ignore
export const commands = [
  CloudRunFlareCommand,
  CloudRunInstrumentCommand,
  CloudRunUninstrumentCommand,
]
