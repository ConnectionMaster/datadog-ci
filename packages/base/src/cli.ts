/* eslint-disable quote-props */
import {CommandClass} from 'clipanion'

import {UploadCommand} from './commands/git-metadata/upload'
import {CheckCommand} from './commands/plugin/check-command'
import {DeployTestsCommand} from './commands/synthetics/deploy-tests-command'
import {ImportTestsCommand} from './commands/synthetics/import-tests-command'
import {RunTestsCommand} from './commands/synthetics/run-tests-command'
import {UploadApplicationCommand} from './commands/synthetics/upload-application-command'

// prettier-ignore
export const commands: Record<string, CommandClass[]> = {
  'git-metadata': [UploadCommand],
  'plugin': [CheckCommand],
  'synthetics': [RunTestsCommand, DeployTestsCommand, UploadApplicationCommand, ImportTestsCommand],
}
