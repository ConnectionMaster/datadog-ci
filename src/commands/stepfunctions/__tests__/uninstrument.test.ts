import {Cli} from 'clipanion/lib/advanced'

import {UninstrumentStepFunctionsCommand} from '../uninstrument'

import {describeStateMachineFixture, subscriptionFilterFixture} from './fixtures/aws-resources'
import {contextFixture, testContext} from './fixtures/cli'

jest.mock('../../../../package.json', () => ({version: '2.0.0'}))

describe('step-functions uninstrument', () => {
  let cli: Cli
  let aws: any
  let changes: any

  beforeAll(() => {
    changes = require('../changes')
    changes.applyChanges = jest.fn().mockImplementation(() => {
      return false
    })

    cli = new Cli()
    cli.register(UninstrumentStepFunctionsCommand)
  })

  let context: testContext
  beforeEach(() => {
    aws = require('../aws')
    context = contextFixture()

    // different function responses may be needed depending on the test
    const stepFunction = describeStateMachineFixture()
    aws.describeStateMachine = jest.fn().mockImplementation(() => stepFunction)

    const subscriptionFilter = subscriptionFilterFixture()
    aws.describeSubscriptionFilters = jest.fn().mockImplementation(() => ({subscriptionFilters: [subscriptionFilter]}))
  })

  describe('paramater validation', () => {
    test('errors if forwarder arn is not set', async () => {
      const exitCode = await cli.run(['step-functions', 'uninstrument'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] --forwarder is required')
    })

    test('errors if forwarder arn is invalid', async () => {
      const exitCode = await cli.run(['step-functions', 'uninstrument', '--forwarder', 'arn:'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] invalid arn format for --forwarder arn:')
    })

    test('removes duplicate step function arns', async () => {
      const exitCode = await cli.run(
        [
          'step-functions',
          'uninstrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(aws.describeStateMachine).toHaveBeenCalledTimes(1)
    })

    test('errors if no step function arn', async () => {
      const exitCode = await cli.run(['step-functions', 'uninstrument'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] must specify at least one --step-function')
    })

    test('errors if any step function arn is invalid', async () => {
      const exitCode = await cli.run(['step-functions', 'uninstrument', '--step-function', 'arn:'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] invalid arn format for --step-function')
    })
  })

  describe('step function logging disabled', () => {
    test('creates step function log group, subscribes it to forwarder, and adds it to step function logging config', async () => {
      const loggingConfiguration = {
        level: 'OFF',
        includeExecutionData: false,
      }
      const stepFunction = describeStateMachineFixture({loggingConfiguration})
      aws.describeStateMachine = jest.fn().mockImplementation(() => stepFunction)

      const exitCode = await cli.run(
        [
          'step-functions',
          'uninstrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] Unable to get Log Group arn from Step Function logging configuration')
    })
  })

  describe('step function logging enabled', () => {
    test('unsubscribes log group in step function logging config from forwarder and removes dd_sls_ci tag', async () => {
      const exitCode = await cli.run(
        [
          'step-functions',
          'uninstrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })
  })

  describe('aws error handling', () => {
    test('errors if unable to fetch step function', async () => {
      aws.describeStateMachine = jest.fn().mockImplementation(() => {
        throw new Error('mock describeStateMachine error')
      })

      const exitCode = await cli.run(
        [
          'step-functions',
          'uninstrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch(
        '[Error] mock describeStateMachine error. Unable to fetch Step Function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction'
      )
    })

    test('errors if unable to fetch step function subscription filters', async () => {
      aws.describeSubscriptionFilters = jest.fn().mockImplementation(() => {
        throw new Error('mock describeSubscriptionFilters error')
      })

      const exitCode = await cli.run(
        [
          'step-functions',
          'uninstrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch(
        '[Error] mock describeSubscriptionFilters error. Unable to fetch Subscription Filter to delete for Log Group /aws/vendedlogs/states/ExampleStepFunction-Logs-test'
      )
    })
  })
})
