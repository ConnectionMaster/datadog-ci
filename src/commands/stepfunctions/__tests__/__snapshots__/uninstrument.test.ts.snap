// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step-functions uninstrument step function logging enabled unsubscribes log group in step function logging config from forwarder and removes dd_sls_ci tag 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
deleteSubscriptionFilter ->
{
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs-test\\"
}
untagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tagKeys\\": [
    \\"dd_sls_ci\\"
  ]
}
"
`;

exports[`stepfunctions uninstrument step function logging enabled unsubscribes log group in step function logging config from forwarder and removes dd_sls_ci tag 1`] = `""`;
